aws-cpp-sdk-core-tests/utils/threading/ExecutorTest.cpp

```diff
@@ -0,0 +1,373 @@
+            if (lastTaskWaiting)
+            {
+                break;
+            }
+        }
+    }
+    taskSignal.notify_all();
+            
+    locker.lock();
+    
+    if (taskOrder.size() != FINAL_SIZE)
+    {
+        doneSignal.wait(locker, [&] {
+            return taskOrder.size() == FINAL_SIZE;
+        });
+    }
+    
+    // Check completion and ordering of tasks.
+    ASSERT_EQ(taskOrder.size(), FINAL_SIZE);
+    ASSERT_EQ(taskOrder.at(3), 3ul);
+    locker.unlock();
+    
+    AWS_END_MEMORY_TEST
+}
+
+/** Tests whether the blocking executor properly delegates tasks concurrently */
+TEST(BlockingExecutorTest, ConcurrencyTimingTest)
+{
+    AWS_BEGIN_MEMORY_TEST(16, 10)
+    
+    const size_t NUM_TASKS = 15;
+    const size_t POOL_SIZE = 5;
+    
+    unsigned int counter = 0;
+    std::mutex counterLock;
+    std::condition_variable doneSignal;
+    
+    auto executeTask = [&] () {
+        std::this_thread::sleep_for(milliseconds(10));
+        std::unique_lock<std::mutex> locker(counterLock);
+        counter++;
+        if (counter == NUM_TASKS)
+        {
+            locker.unlock();
+            doneSignal.notify_all();
+        }
+    };
+
+    // Time how long it takes to run the tasks with an underlying executor that's concurrent
+    std::shared_ptr<DefaultExecutor> mockExecutor =
+            Aws::MakeShared<DefaultExecutor>(EXECUTOR_TEST_ALLOCATION_TAG);
+    BlockingExecutor blockingExecutor(mockExecutor, POOL_SIZE);
+    
+    high_resolution_clock::time_point startTime = high_resolution_clock::now();
+    for (u_long i = 0; i < NUM_TASKS; i++)
+    {
+        blockingExecutor.Submit(executeTask);
+    }
+    std::unique_lock<std::mutex> locker(counterLock);
+    while (counter != NUM_TASKS)
+    {
+        doneSignal.wait(locker);
+    }
+    counter = 0;
+    locker.unlock();
+    high_resolution_clock::time_point endTime = high_resolution_clock::now();
+    auto asyncDuration = duration_cast<milliseconds>(endTime - startTime).count();
+    
+    // Time how long it takes to run the tasks with an underlying executor that doesn't use threads
+    std::shared_ptr<ThreadlessMockExecutor> threadlessMockExecutor =
+            Aws::MakeShared<ThreadlessMockExecutor>(EXECUTOR_TEST_ALLOCATION_TAG);
+    BlockingExecutor threadlessBlockingExecutor(threadlessMockExecutor, POOL_SIZE);
+
+    startTime = high_resolution_clock::now();
+    for (u_long i = 0; i < NUM_TASKS; i++)
+    {
+        threadlessBlockingExecutor.Submit(executeTask);
+    }
+    locker.lock();
+    while (counter != NUM_TASKS)
+    {
+        doneSignal.wait(locker);
+    }
+    endTime = high_resolution_clock::now();
+    auto sequentialDuration = duration_cast<milliseconds>(endTime - startTime).count();
+    
+    /* Check if the concurrent time was faster, with an adjustable tolerance for the strictness of
+     * the test */
+    const double TOLERANCE = 1.0;
+    ASSERT_LE(asyncDuration, sequentialDuration * TOLERANCE);
```

Comment:
This seems non-deterministic to me. For example, a run loop with async IO will likely beat out multiple threads with synchronous IO. Also, who knows if the operating system thread creation time is faster than running 15 stack frames or not? I would certainly imagine that this test would fail on an embedded system.